File: lib/agents/metricsAgent.ts
import { HIGH_COMPLIANCE_MARKETS } from "@/lib/markets";
import type {
  Citation,
  DocumentMeta,
  DocumentRelevance,
  DocumentSection,
  ReportMetrics,
  RiskItem,
  WorkflowStep,
} from "@/lib/types";
import type { PersonaKey } from "./personaMapperAgent";

const clamp = (v: number, min: number, max: number) => Math.max(min, Math.min(max, v));

interface MetricsInput {
  documents: DocumentMeta[];
  sections: DocumentSection[];
  citations: Citation[];
  risks: RiskItem[];
  workflowSteps: WorkflowStep[];
  market: string;
  personaKey: PersonaKey;
  personaTitle: string;
}

function complianceSensitivity(market: string, personaKey: PersonaKey): ReportMetrics["complianceSensitivity"] {
  if (HIGH_COMPLIANCE_MARKETS.includes(market) || market === "EMEA") return "High";
  if (personaKey === "compliance") return "High";
  return "Moderate";
}

function documentRelevance(documents: DocumentMeta[], personaTitle: string): DocumentRelevance[] {
  return documents
    .map((d, i) => {
      const match = d.usedByPersonas.includes(personaTitle);
      const score = clamp(72 + (match ? 20 : 6) - i * 2, 60, 98);
      return { document: d.short, score };
    })
    .sort((a, b) => b.score - a.score)
    .slice(0, 6);
}

/**
 * Metrics Agent.
 * Produces deterministic, input-driven report metrics. More selected documents
 * generally raise confidence and citation coverage.
 */
export function metricsAgent(input: MetricsInput): ReportMetrics {
  const documentsAnalyzed = input.documents.length;
  const passagesIndexed = input.sections.length * 41;
  const coverageRatio = documentsAnalyzed / 8;

  const confidence = clamp(Math.round(82 + coverageRatio * 14), 82, 96);
  const personaRelevanceScore = clamp(76 + documentsAnalyzed * 2 + 4, 70, 95);
  const workflowCompleteness = clamp(80 + input.workflowSteps.length * 4, 80, 100);
  const citationCoverage = clamp(70 + documentsAnalyzed * 3, 70, 95);

  return {
    documentsAnalyzed,
    passagesIndexed,
    citationsUsed: input.citations.length,
    risksFlagged: input.risks.length,
    confidence,
    personaRelevanceScore,
    workflowCompleteness,
    citationCoverage,
    complianceSensitivity: complianceSensitivity(input.market, input.personaKey),
    documentRelevance: documentRelevance(input.documents, input.personaTitle),
  };
}

File: backend/app/agents/metrics_agent.py
"""Metrics agent — deterministic, input-driven report metrics."""

from __future__ import annotations

from typing import Any, Dict, List

from app.tools.scoring_tools import (
    compute_citation_coverage,
    compute_compliance_sensitivity,
    compute_confidence,
    compute_document_relevance,
    compute_persona_relevance,
    compute_workflow_completeness,
)


def metrics_agent(
    documents: List[Dict[str, Any]],
    sections: List[Dict[str, Any]],
    citations: List[Dict[str, Any]],
    risks: List[Dict[str, Any]],
    workflow_steps: List[Dict[str, Any]],
    market: str,
    persona_key: str,
    persona_title: str,
) -> Dict[str, Any]:
    documents_analyzed = len(documents)
    return {
        "documentsAnalyzed": documents_analyzed,
        "passagesIndexed": len(sections) * 41,
        "citationsUsed": len(citations),
        "risksFlagged": len(risks),
        "confidence": compute_confidence(documents_analyzed),
        "personaRelevanceScore": compute_persona_relevance(documents_analyzed),
        "workflowCompleteness": compute_workflow_completeness(len(workflow_steps)),
        "citationCoverage": compute_citation_coverage(documents_analyzed),
        "complianceSensitivity": compute_compliance_sensitivity(market, persona_key),
        "documentRelevance": compute_document_relevance(documents, persona_title),
    }
